python/morpheus/morpheus/utils/shared_process_pool.py

```diff
@@ -0,0 +1,409 @@
+
+        new_total_usage = self._total_usage - self._stage_usage.get(stage_name, 0.0) + percentage
+
+        if new_total_usage > 1.0:
+            raise ValueError("Total percentage cannot exceed 1.0.")
+
+        self._stage_usage[stage_name] = percentage
+        self._total_usage = new_total_usage
+
+        allowed_processes_num = max(1, int(self._total_max_workers * percentage))
+        self._stage_semaphores[stage_name] = self._manager.Semaphore(allowed_processes_num)
+
+        if stage_name not in self._task_queues:
+            self._task_queues[stage_name] = self._manager.Queue()
+
+        logger.debug("SharedProcessPool.set_usage(): stage_usage: %s", self._stage_usage)
+        logger.debug("SharedProcessPool.set_usage(): stage semaphores: %s", allowed_processes_num)
+
+    def start(self):
+        """
+        Start the SharedProcessPool that is currently stopped and keep the settings before last shutdown.
+
+        Raises
+        ------
+        RuntimeError
+            If the SharedProcessPool is not shutdown.
+        """
+        if self._status != PoolStatus.SHUTDOWN:
+            logger.warning("SharedProcessPool.start(): Cannot start a SharedProcessPool that is not shutdown.")
+            return
+
+        self._launch_workers()
+        self._status = PoolStatus.RUNNING
+
+    def reset(self):
+        """
+        Clear all the previous settings and restart the SharedProcessPool.
+
+        Raises
+        ------
+        RuntimeError
+            If the SharedProcessPool is not already shutdown.
+        """
+        if self._status != PoolStatus.SHUTDOWN:
+            raise RuntimeError(
+                "SharedProcessPool.reset(): Cannot reset a SharedProcessPool that is not already shutdown.")
+
+        self._initialize()
+
+    def stop(self):
+        """
+        Complete existing tasks and stop the SharedProcessPool.
+        """
+        if self._status not in (PoolStatus.RUNNING, PoolStatus.INITIALIZING):
+            logger.warning("SharedProcessPool.stop(): Cannot stop a SharedProcessPool that is not running.")
+            return
+
+        # no new tasks will be accepted from this point
+        self._status = PoolStatus.STOPPING
+
+        # wait for all task queues to be empty
+        task_queue_count = len(self._task_queues)
+        empty_task_queues = set()
+        while len(empty_task_queues) < task_queue_count:
+            for stage_name, task_queue in self._task_queues.items():
+                if task_queue.empty():
+                    empty_task_queues.add(stage_name)
+
+        self._cancellation_token.cancel()
+
+        for i, p in enumerate(self._processes):
+            p.join()
+            logger.debug("Process %s/%s has been joined.", i + 1, self._total_max_workers)
+
+        logger.debug("SharedProcessPool.stop(): All tasks have been completed. SharedProcessPool has been stopped.")
+        self._status = PoolStatus.SHUTDOWN
```

Comment:
Stop should be non-blocking. `stop()` means, do not accept any more work, let everything finish naturally. `join()` is the blocking function which means, wait unti everything is cleaned up and complete.